### ISSUE TITLE
为什么不能做环线

### ISSUE THREAD

**Issue by NiuBoss123** [wongchito/RailMapGenerator#24]
目前，很多城市开始规划环线，令我失望的是这个编辑器不能做环线，当我在Ai上做着线路图时，我想到了你们，希望可以给你们一些建议。
### 建议
环线有可能有支线吗，答案是肯定的，以首尔地铁来说，2号线就有2条支线，希望在更新环线时可以添加这个功能。
环线是要有方向的，这一点只需要加上箭头即可
环线有终点吗，有，环线的终点需要分出来一个版面，因为部分城市把环线的终点给分出一个站台以供上下终点客流，这一点要注意。

**Comment by NiuBoss** [comment]
At present, many cities are beginning to plan circular lines. I am disappointed that this editor cannot be used for circular lines. When I was making a road map on Ai, I thought of you and hoped that I could give you some suggestions.

Suggest
Is there a branch line in the ring line? The answer is yes. As far as the Seoul Metro is concerned, there are two branch lines on line 2. I hope that this function can be added when the ring line is updated.
The loop line must have a direction, and you only need to add an arrow at this point.
Does the ring line have an end point? Yes, the end point of the ring line needs to be divided into a layout, because some cities divide the end point of the ring line to a platform for passenger flow to and from the end point. It should be noted that

**Comment by wongchito** [comment]
Thanks for getting in touch. Circle line (loop line) is indeed frequently requested by users, but there're a few concerns before implementing this feature. 

1. RMG currently supports 3 styles, where Shanghai Metro is the only system containing a circle line - Line 4. However, Line 4 shares parts of its stations with Line 3 and both lines have to be drawn together. We're still investigating the possibility of introducing additional lines into the canvas within the existing data structure (directed acyclic graph stored as a chained list). 

2. Line 11 of Guangzhou Metro is currently under construction. We do have a plan for supporting it but we're unable to predict the style Guangzhou Metro chose to use. We won't implement anything that is not exist in or can't be analogously inferred from the official style. 

3. There's no plan of constructing a circle line in Hong Kong (MTR). 

4. Except for those are described in 1, one other difficulty is the algorithm for deciding the position (both horizontal and vertical) of a station on the canvas. RMG doesn't allow user to choose a station's position but compute them with a combination of algorithms from the field of graph theory. Adding branches for circle line may hugely increase the complexity of the algorithms. 

Long story short, there're mainly 2 difficulties: **data structure** and **algorithms for computing station's position**. If you have any idea solving these issues, please feel free to contact us. Thank you. 

長話短說，現主要有兩點困難：**數據結構**及**計算車站位置的演算法**。如閣下有任何改善上述問題的想法，請與我們聯絡。多謝。

**Comment by thekingofcity** [comment]
As the author of shmetro, I'm interested in 

>>  We're still investigating the possibility of introducing additional lines into the canvas within the existing data structure (directed acyclic graph stored as a chained list).

and glad to be in part of the discussion (of course with sharing lines possible).

PS: maybe it's time to list a to-do list and link them to some milestones?

**Comment by maliaoshui** [comment]
`yujs`
wongduoyu
<hfk>
  <htm >

**Comment by maliaoshui** [comment]
`<html>
  {home page
     <css>
<wongduoyu>
}`

**Comment by wongchito** [comment]
> Only one child is allowed for station `linestart` and one parent for `lineend`. For circle line, this limit should raised to 2. However, this change will break the current algorithm of computing possible location of station to be added. 

> Branches in circle line will increase the difficulty of the algorithm implementation. 

For phase 1 solution, circle line can only be formed from a normal line without branches. Station adding function, reversing line function (etc.) will be completely disabled for circle line. 

Circle line can be de-looped and be reverted to normal line, in which all function disabled above will be re-enabled. 

Works required:

- [ ] `isCircleLine(): boolean`
- [ ] `drawLine()` to support circle line
- [ ] `connectLineStartAndLineEnd()`
- [ ] `deloop(stn1, stn2)`

**Comment by thekingofcity** [comment]
So I'm curious about why we need to disable the station adding function and the reversing line function.

By disabling the station adding function, the whole circle line feature would be more like a one-time presentation feature rather than a general use feature.

And by disabling the reversing line function, a user would not be able to see the reversed `... A B C ...`, even if he or she change the direction. So all he or she can do is to deloop the line and reverse and then connect?

I'm really look forward to the circle line feature, but from a user perspective the first stage solution seems a bit confusing.

**Comment by wongchito** [comment]
It's about "go faster and break less". 

Adding circle line feature is a really big change. We want to split it into multiple small deliverable, to avoid breaking existing flow. We certainly don't want user to use some functions that are not fully tested and lost their valuable data, but we are not able to deliver this huge change in one shot and promise everything is working as expected. 

For the reasons above, I tend to split this feature into several phases:

1. Update draw line function such that RMG can render a param config with circle line settings correctly. At the same time, we will not provide adding station or reversing line function for circle line setup. (View only)
2. Re-implement algorithm of adding station, to support normal line with or without branches, and circle line. Implement all other disabled functions for circle line. 
3. Toggle on features above step by step. 

This discussion is open and I'm happy to re-design the road map of this feature should you have any advice.

**Comment by thekingofcity** [comment]
That made me clear, thanks for the explanation.

Since I suggested some changes on the underlying data structure in [#41](https://github.com/wongchito/RailMapGenerator/issues/41#issuecomment-882079777), I wondered whether the circle line feature will work on these changes or not :)

**Comment by thekingofcity** [comment]
Hi, this issue is selected as our core feature for the next generation. We'll work on this as soon as poosible.

**Comment by thekingofcity** [comment]
As we are in the very early stage of the loop line, I want to share some thoughts on the internal data structure.

> Only one child is allowed for station linestart and one parent for lineend. For circle line, this limit should raised to 2. However, this change will break the current algorithm of computing possible location of station to be added.

This is of course a valid option, but others include:

* Remove the `linestart` and `lineend` nodes entirely.
  pros: cyclical list can be done in a textbook approach, which should be friendly to other contributors
  cons: break the current algorithm like `drawLine` and `getXShareMTR`, and the start/end direction can be solved via another `start: string[]` and `end: string[]` field.
* Change loop line to an entire visual feature by drawing a loop path when the first and the last station share the same name.
  pros: almost no breaks in the development
  cons: 💩 on codes and faces problems when circle line have branches

All these changes can be done in a "go faster and break less" manner with functions like `isCircleLine(): boolean`.

**Comment by thekingofcity** [comment]
Loop line will be implemented in these following issues.

* Railmap and indoor canvas #266 
* Destination canvas
* UI and redux
* Maybe some tests?

**Comment by thekingofcity** [comment]
Loop line if finally here. Please join me to celebrate this moment. 🎉

![rmg y6id 塘桥 Tangqiao l](https://user-images.githubusercontent.com/3353040/158186097-c4ff5321-99dc-4361-8734-5a313e7df737.png)
> Railmap canvas with left direction

![rmg y6id 塘桥 Tangqiao r](https://user-images.githubusercontent.com/3353040/158186120-c412caf6-d1ab-4836-93c8-6d3e1cf28369.png)
> Railmap canvas with right direction

![rmg y6id 塘桥 Tangqiao indoor](https://user-images.githubusercontent.com/3353040/158185653-62f2f96f-cdc2-416b-a8f9-4b5d520f97ff.png)
> Indoor canvas

![rmg hhar 鲁班路 Luban_Road](https://user-images.githubusercontent.com/3353040/158186149-d758a8cb-a79c-45f7-a284-6324120810e4.png)
> Railmap canvas with auto-rotation for different stations
